packages/csvviewer/src/model.ts

```diff
@@ -0,0 +1,663 @@
+        rowDelimiter = '\r\n';
+      } else {
+        rowDelimiter = '\r';
+      }
+    }
+    this._rowDelimiter = rowDelimiter;
+
+    if (quoteParser === undefined) {
+      // Check for the existence of quotes if the quoteParser is not set.
+      quoteParser = (data.indexOf(quote) >= 0);
+    }
+    this._parser = quoteParser ? 'quotes' : 'noquotes';
+
+    // Parse the data.
+    this._parseAsync();
+
+    // Cache the header row.
+    if (header === true && this._columnCount > 0) {
+      let h = [];
+      for (let c = 0; c < this._columnCount; c++) {
+        h.push(this._getField(0, c));
+      }
+      this._header = h;
+    }
+  }
+
+  /**
+   * Whether this model has been disposed.
+   */
+  get isDisposed() {
+    return this._isDisposed;
+  }
+
+  /**
+   * A promise that resolves when the model has parsed all of its data.
+   */
+  get ready() {
+    return this._ready.promise;
+  }
+
+  /**
+   * Get the row count for a region in the data model.
+   *
+   * @param region - The row region of interest.
+   *
+   * @returns - The row count for the region.
+   */
+  rowCount(region: DataModel.RowRegion): number {
+    if (region === 'body') {
+      if (this._header.length === 0) {
+        return this._rowCount;
+      } else {
+        return this._rowCount - 1;
+      }
+    }
+    return 1;
+  }
+
+  /**
+   * Get the column count for a region in the data model.
+   *
+   * @param region - The column region of interest.
+   *
+   * @returns - The column count for the region.
+   */
+  columnCount(region: DataModel.ColumnRegion): number {
+    if (region === 'body') {
+      return this._columnCount;
+    }
+    return 1;
+  }
+
+  /**
+   * Get the data value for a cell in the data model.
+   *
+   * @param region - The cell region of interest.
+   *
+   * @param row - The row index of the cell of interest.
+   *
+   * @param column - The column index of the cell of interest.
+   *
+   * @param returns - The data value for the specified cell.
+   */
+  data(region: DataModel.CellRegion, row: number, column: number): any {
+    let value: string;
+
+    // Look up the field and value for the region.
+    switch (region) {
+    case 'body':
+      if (this._header.length === 0) {
+        value = this._getField(row, column);
+      } else {
+        value = this._getField(row + 1, column);
+      }
+      break;
+    case 'column-header':
+      if (this._header.length === 0) {
+        value = (column + 1).toString();
+      } else {
+        value = this._header[column];
+      }
+      break;
+    case 'row-header':
+      value = (row + 1).toString();
+      break;
+    case 'corner-header':
+      value = '';
+      break;
+    default:
+      throw 'unreachable';
+    }
+
+    // Return the final value.
+    return value;
+  }
+
+  /**
+   * Dispose the resources held by this model.
+   */
+  dispose() {
+    if (this._isDisposed) {
+      return;
+    }
+    this._resetParser();
+    this._data = null;
```

Comment:
Why make this disposable? Also, resetting the parser causes the changed signal to be emitted, which means listeners will be dealing with a "disposed" object.

Reply:
I made it disposable because it has async parsing that could be scheduled. Calling the dispose method explicitly cancels any pending parsing work.

Reply:
Again, good catch. I tried to factor out the code that cancels the async processing, but that introduced this problem. I'll make another commit separating these two usecases.